### DISCUSSION TITLE
How do we use shell completion without installing anything

### DISCUSSION THREAD

**Discussion by Hubro** [pallets/click#2323] [Q&A]
I'm making a utility script for one of our company's repositories called `scripts/dev.py`, with an alias `alias dev=./scripts/dev.py`.

For convenience, I want to enable shell autocompletion, but [the instructions](https://click.palletsprojects.com/en/8.1.x/shell-completion/) say "Completion is only available if a script is installed and invoked through an entry point".

Why is this a limitation? Is there any way to work around it?

I've tried to invoke our script with `_DEV_COMPLETE=zsh_source dev` but the environment variable is ignored.

**Comment by davidism** [comment]
You can't. This is a limitation of how shell completion works in shells, it's not something Click can control.

**Reply by Hubro** [reply]
I got it working by renaming the script to just `dev` (with an appropriate shebang) and adding the `scripts` directory to PATH, rather than using an alias. So, strictly, the script doesn't actually need to be "installed", depending on what is meant by that. I guess the precise limitations could be documented a little better, it had me believing I needed an install script and stuff :smile:

**Reply by davidism** [reply]
That has nothing to do with Python at that point, but you messing with your shell's path and what it sees as a binary. You might as well have installed the package at that point, you would have got an isolated venv, dependency management, and path management for free. We want users to use Python's standard packaging workflows because otherwise they'll just reinvent them worse.

**Reply by Hubro** [reply]
This script is basically a developement utility for a single repository, we want our developers to be able to check out the code and use the script immediately. Adding an installer for it seems like overkill, especially since it's not usable for anything outside of this specific repository.

I guess my use case might be unusual.
